File: C++/Graphs/35_FarFromLandPossible.cpp
int bfs(vector<vector<int>> &grid, queue<pair<int, int>> &q, int n)
{
    int maxDist = 0;
    vector<vector<int>> dir{{-1, 0}, {1, 0}, {0, 1}, {0, -1}};
    while (!q.empty())
    {
        auto p = q.front();
        int x = p.first;
        int y = p.second;
        q.pop();
        for (int k = 0; k < 4; k++)
        {
            int a = x + dir[k][0];
            int b = y + dir[k][1];
            if (a >= 0 and b >= 0 and a < n and b < n and grid[a][b] == 0)
            {
                grid[a][b] = grid[x][y] + 1;
                maxDist = max(maxDist, grid[a][b]);
                q.push({a, b});
            }
        }
    }
    return maxDist - 1;
}

int maxDistance(vector<vector<int>> &grid)
{
    int n = grid.size();
    queue<pair<int, int>> q;

    for (int i = 0; i < n; ++i)
    {
        for (int j = 0; j < n; ++j)
        {
            if (grid[i][j] == 1)
            {
                q.push({i, j});
            }
        }
    }

    return bfs(grid, q, n);
}

File: Java/Graphs/35_FarFromLandPossible.java
class Solution {
    public int maxDistance(int[][] grid) {
        int n = grid.length;
        int m = grid[0].length;
        int dir[][] = {{-1,0},{0,-1},{1,0},{0,1}};
        Queue<Integer> que = new LinkedList<>();
        int c1 = 0, c0 = 0;
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j] == 1){
                    que.add(i*m + j);
                    c1++;
                }
                else c0++;
            }
        }
        if(c1 == 0 || c0 == 0) return -1;
        int level = 0;
        while(que.size() != 0){
            int size = que.size();
            while(size-->0){
                int idx = que.remove();
                int i = idx/m;
                int j = idx%m;
                for(int d=0;d<4;d++){
                    int r = i + dir[d][0];
                    int c = j + dir[d][1];
                    if(r>=0 && c>=0 && r<n && c<m && grid[r][c] == 0){
                        grid[r][c] = 1;
                        que.add(r*m + c);
                    }
                }
            }
            level++;
        }
        
        return level-1;
    }
}